File: 06-observability/04-cardinality/golang/cardinality_overflow.go
// Layer 6 Topic 4 - One label, two failures, and the three places you can fix it.
//
// Why Go: the mechanism is a spec-mandated SDK default, not a runtime property,
// so this file is not here to be different arithmetic. It is here because the
// CONFIGURATION SURFACE is different, and the difference matters when you are
// trying to fix this at 3am:
//
//	Python  a per-view setting on the MeterProvider, or the process-wide
//	        OTEL_METRIC_CARDINALITY_LIMIT env var.
//	Go      metric.NewView(...) -> a function from Instrument to Stream, so the
//	        limit, the attribute FILTER and the aggregation are all decided in
//	        one place, by code, per instrument.
//
// That second option -- Stream.AttributeFilter -- is the lever Python does not
// give you as directly, and it is the better fix: dropping the attribute keeps
// your ratios honest, where letting it overflow does not. Both are modelled
// below, side by side, against the same traffic.
//
// No OpenTelemetry SDK is installed on this machine, so the counter, the view,
// the cardinality limit and the overflow attribute are written out here in
// about eighty lines of standard library. They implement the spec's rule: past
// the limit, the measurement is still counted, but its attributes are replaced
// by {otel.metric.overflow: true}.
//
// What to look for in the output
// ------------------------------
//  1. The three views' series counts, side by side, against the same 200k
//     requests. Same data, same instrument, three configurations.
//  2. The `sum by (customer_id)` gap under the limit -- the silent failure.
//  3. Section 4's MEASURED bytes per series, from runtime.MemStats on this
//     machine, then multiplied out. The measurement is real; the multiplication
//     is labelled as a multiplication.
//  4. Section 5, the collector fix: when thirty services are emitting a bad
//     label you cannot redeploy thirty services, but you can edit one config.
//     The last block shows the same traffic passing through a drop rule and a
//     hash rule, and what each one costs you.
package main

import (
	"fmt"
	"math/rand"
	"runtime"
	"strings"
)

// The spec default. In the real Go SDK this is
// metric.NewView(metric.Instrument{Name: "..."},
//
//	metric.Stream{AggregationCardinalityLimit: 2000})
const defaultCardinalityLimit = 2000

// ---------------------------------------------------------------------------
// A counter with a view: a cardinality limit and an optional attribute filter.
// ---------------------------------------------------------------------------

type attrSet struct {
	route      string
	method     string
	status     int
	customerID string // "" means the attribute is not present
	overflow   bool
}

func (a attrSet) key() string {
	if a.overflow {
		return "otel.metric.overflow=true"
	}
	return fmt.Sprintf("route=%s,method=%s,status=%d,customer_id=%s",
		a.route, a.method, a.status, a.customerID)
}

// Stream is the Go SDK's view output, cut down to the two fields this topic is
// about. AttributeFilter runs BEFORE the cardinality limit, which is the whole
// reason it is the better fix.
type Stream struct {
	Name                        string
	AggregationCardinalityLimit int                   // 0 = unlimited
	AttributeFilter             func(attrSet) attrSet // nil = keep everything
}

type Counter struct {
	stream           Stream
	series           map[string]int64
	attrs            map[string]attrSet
	overflowedPoints int64
	rejectedAttrSets map[string]struct{}
}

func NewCounter(stream Stream) *Counter {
	return &Counter{
		stream:           stream,
		series:           make(map[string]int64),
		attrs:            make(map[string]attrSet),
		rejectedAttrSets: make(map[string]struct{}),
	}
}

func (c *Counter) Add(value int64, a attrSet) {
	if c.stream.AttributeFilter != nil {
		a = c.stream.AttributeFilter(a)
	}
	k := a.key()
	if _, ok := c.series[k]; ok {
		c.series[k] += value
		return
	}
	limit := c.stream.AggregationCardinalityLimit
	// The spec reserves one slot for the overflow series itself.
	if limit == 0 || len(c.series) < limit-1 {
		c.series[k] = value
		c.attrs[k] = a
		return
	}
	overflow := attrSet{overflow: true}
	ok := overflow.key()
	c.series[ok] += value
	c.attrs[ok] = overflow
	c.overflowedPoints += value
	c.rejectedAttrSets[k] = struct{}{}
}

// Total is sum(rate(...)) with no label matcher. Always correct.
func (c *Counter) Total() int64 {
	var total int64
	for _, v := range c.series {
		total += v
	}
	return total
}

// SumBy is sum by (label) (...). The overflow point has no such label, so it
// silently drops out.
func (c *Counter) SumBy(label string) map[string]int64 {
	out := make(map[string]int64)
	for k, v := range c.series {
		a := c.attrs[k]
		if a.overflow {
			continue
		}
		switch label {
		case "customer_id":
			if a.customerID == "" {
				continue
			}
			out[a.customerID] += v
		case "route":
			out[a.route] += v
		default:
			panic("unknown label " + label)
		}
	}
	return out
}

// SumWhere is any query with a label matcher, e.g. {status=~"5.."}. Same trap:
// the overflow point carries no status either.
func (c *Counter) SumWhere(pred func(attrSet) bool) int64 {
	var total int64
	for k, v := range c.series {
		a := c.attrs[k]
		if a.overflow {
			continue
		}
		if pred(a) {
			total += v
		}
	}
	return total
}

func (c *Counter) ActiveSeries() int { return len(c.series) }

// ---------------------------------------------------------------------------
// Traffic
// ---------------------------------------------------------------------------

var routes = buildRoutes()
var methods = []string{"GET", "POST", "PUT", "DELETE", "PATCH"}
var statuses = []int{200, 201, 204, 400, 401, 404, 429, 500}
var statusWeights = []int{70, 8, 5, 5, 3, 4, 3, 2}

func buildRoutes() []string {
	// 40 templated routes. The raw-path version of this list is unbounded,
	// because `?utm_source=` is user input.
	out := []string{"/health", "/ready", "/metrics", "/checkout",
		"/pricing/quote", "/search", "/login", "/logout"}
	for _, r := range []string{"orders", "customers", "items", "invoices",
		"shipments", "returns", "quotes", "carts"} {
		out = append(out,
			"/"+r,
			"/"+r+"/{id}",
			"/"+r+"/{id}/events",
			"/"+r+"/{id}/audit")
	}
	return out
}

const (
	customers = 10_000
	requests  = 200_000
)

func weightedPick(rng *rand.Rand, weights []int) int {
	total := 0
	for _, w := range weights {
		total += w
	}
	n := rng.Intn(total)
	for i, w := range weights {
		n -= w
		if n < 0 {
			return i
		}
	}
	return len(weights) - 1
}

func generateTraffic(rng *rand.Rand) []attrSet {
	out := make([]attrSet, 0, requests)
	for i := 0; i < requests; i++ {
		out = append(out, attrSet{
			route:      routes[rng.Intn(len(routes))],
			method:     methods[weightedPick(rng, []int{70, 20, 5, 3, 2})],
			status:     statuses[weightedPick(rng, statusWeights)],
			customerID: fmt.Sprintf("cust-%05d", rng.Intn(customers)),
		})
	}
	return out
}

func section(title string) {
	fmt.Printf("\n%s\n%s\n", title, strings.Repeat("-", 52))
}

func main() {
	rng := rand.New(rand.NewSource(20260818))

	fmt.Println("Layer 6 Topic 4 - cardinality in Go: three views over one instrument")
	fmt.Printf("%s   spec default AggregationCardinalityLimit = %d\n",
		runtime.Version(), defaultCardinalityLimit)
	fmt.Println(strings.Repeat("=", 72))

	traffic := generateTraffic(rng)

	// -----------------------------------------------------------------------
	section("1. Series count is a product, and you can do it on a napkin")
	bounded := len(routes) * len(methods) * len(statuses)
	fmt.Printf("  routes x methods x statuses   %d x %d x %d = %s series\n",
		len(routes), len(methods), len(statuses), comma(int64(bounded)))
	fmt.Printf("  ... x customer_id             %s x %s = %s series\n",
		comma(int64(bounded)), comma(customers), comma(int64(bounded)*customers))
	fmt.Println()
	fmt.Println("  Nothing about the recording changed. One label multiplied.")

	// -----------------------------------------------------------------------
	section("2. Three views, one instrument, the same 200,000 requests")

	// View A: the label is not there at all. This is what you shipped last week.
	viewBounded := NewCounter(Stream{
		Name:                        "http.server.requests",
		AggregationCardinalityLimit: defaultCardinalityLimit,
		AttributeFilter: func(a attrSet) attrSet {
			a.customerID = ""
			return a
		},
	})

	// View B: customer_id added, default limit left in place. The silent one.
	viewLimited := NewCounter(Stream{
		Name:                        "http.server.requests",
		AggregationCardinalityLimit: defaultCardinalityLimit,
	})

	// View C: customer_id added, limit removed. The loud one -- do not do this
	// in a process that shares a TSDB with anything you care about.
	viewUnlimited := NewCounter(Stream{
		Name:                        "http.server.requests",
		AggregationCardinalityLimit: 0,
	})

	for _, a := range traffic {
		viewBounded.Add(1, a)
		viewLimited.Add(1, a)
		viewUnlimited.Add(1, a)
	}

	fmt.Printf("  %-34s %12s %14s\n", "view", "series", "sum(rate(...))")
	fmt.Printf("  %-34s %12s %14s\n", strings.Repeat("-", 34),
		strings.Repeat("-", 12), strings.Repeat("-", 14))
	fmt.Printf("  %-34s %12s %14s\n", "A: AttributeFilter drops it",
		comma(int64(viewBounded.ActiveSeries())), comma(viewBounded.Total()))
	fmt.Printf("  %-34s %12s %14s\n", "B: limit 2000 (the default)",
		comma(int64(viewLimited.ActiveSeries())), comma(viewLimited.Total()))
	fmt.Printf("  %-34s %12s %14s\n", "C: limit 0 (unlimited)",
		comma(int64(viewUnlimited.ActiveSeries())), comma(viewUnlimited.Total()))
	fmt.Println()
	fmt.Println("  All three totals agree. That is the point: the total is the one")
	fmt.Println("  number that survives every version of this failure, which is")
	fmt.Println("  exactly why the total is not where you will notice it.")

	// -----------------------------------------------------------------------
	section("3. The silent failure, in the two queries that disagree")

	total := viewLimited.Total()
	byCustomer := viewLimited.SumBy("customer_id")
	var visible int64
	for _, v := range byCustomer {
		visible += v
	}
	gap := total - visible

	trueErrors := int64(0)
	for _, a := range traffic {
		if a.status >= 500 {
			trueErrors++
		}
	}
	metricErrors := viewLimited.SumWhere(func(a attrSet) bool { return a.status >= 500 })

	fmt.Printf("  sum(rate(...))                       %14s\n", comma(total))
	fmt.Printf("  sum by (customer_id) (rate(...))     %14s\n", comma(visible))
	fmt.Printf("  gap                                  %14s  (%.1f%% of traffic)\n",
		comma(gap), 100*float64(gap)/float64(total))
	fmt.Printf("  customer_id values surviving         %14s of %s\n",
		comma(int64(len(byCustomer))), comma(customers))
	fmt.Printf("  attribute sets rejected              %14s\n",
		comma(int64(len(viewLimited.rejectedAttrSets))))
	fmt.Println()
	fmt.Printf("  error ratio, ground truth            %13.2f%%\n",
		100*float64(trueErrors)/float64(requests))
	fmt.Printf("  error ratio, as the metric reports   %13.2f%%  <- your alert\n",
		100*float64(metricErrors)/float64(total))
	fmt.Println()
	fmt.Println("  The overflow point carries the count and NOTHING else -- no")
	fmt.Println("  status, no route, no method. So every query with a label matcher")
	fmt.Println("  loses it, and a ratio whose numerator filters and whose")
	fmt.Println("  denominator does not is wrong in the safe-looking direction.")

	// -----------------------------------------------------------------------
	section("4. The loud failure: bytes per series, measured then multiplied")

	perSeries10k := measureBytesPerSeries(10_000)
	perSeries100k := measureBytesPerSeries(100_000)

	fmt.Println("  measured on THIS machine with runtime.ReadMemStats:")
	fmt.Printf("    %8s series   %6.0f bytes/series\n", comma(10_000), perSeries10k)
	fmt.Printf("    %8s series   %6.0f bytes/series\n", comma(100_000), perSeries100k)
	fmt.Println()
	projected := int64(bounded) * customers
	fmt.Println("  extrapolated (measured cost x series count -- NOT a measurement):")
	fmt.Printf("    %8s series   %6.1f GB in this toy store alone\n",
		comma(projected), float64(projected)*perSeries100k/1e9)
	fmt.Println()
	fmt.Println("  Prometheus pays more per series than a Go map does: head block,")
	fmt.Println("  inverted index, label-value dictionary, and query fan-out all")
	fmt.Println("  scale with the count. Take the SHAPE from this number -- a")
	fmt.Println("  straight line with your customer count on the x axis -- and get")
	fmt.Println("  the magnitude from prometheus_tsdb_head_series on the real thing.")

	// -----------------------------------------------------------------------
	section("5. The collector fix: one config instead of thirty deploys")

	// A collector processor is a function from attributes to attributes,
	// applied to every service's telemetry at once. That is the entire reason
	// it is the only intervention available during an incident.
	drop := func(a attrSet) attrSet { a.customerID = ""; return a }
	hash := func(a attrSet) attrSet {
		// attributes/transform can hash instead of dropping: you keep a bounded
		// number of buckets and lose the ability to name one customer.
		h := uint32(2166136261)
		for i := 0; i < len(a.customerID); i++ {
			h = (h ^ uint32(a.customerID[i])) * 16777619
		}
		a.customerID = fmt.Sprintf("bucket-%02d", h%16)
		return a
	}

	dropped := NewCounter(Stream{Name: "http.server.requests",
		AggregationCardinalityLimit: defaultCardinalityLimit, AttributeFilter: drop})
	hashed := NewCounter(Stream{Name: "http.server.requests",
		AggregationCardinalityLimit: defaultCardinalityLimit, AttributeFilter: hash})
	// Same hash, with the limit raised deliberately to fit the series it
	// produces. Raising the limit is only safe when you have done the
	// multiplication first -- which is the whole habit this topic is teaching.
	hashedRaised := NewCounter(Stream{Name: "http.server.requests",
		AggregationCardinalityLimit: 50_000, AttributeFilter: hash})
	for _, a := range traffic {
		dropped.Add(1, a)
		hashed.Add(1, a)
		hashedRaised.Add(1, a)
	}

	trueRatio := 100 * float64(trueErrors) / float64(requests)
	errRatio := func(c *Counter) float64 {
		e := c.SumWhere(func(a attrSet) bool { return a.status >= 500 })
		return 100 * float64(e) / float64(c.Total())
	}

	fmt.Printf("  %-30s %9s %9s  %s\n", "collector rule", "series", "error %",
		"can you still name a customer?")
	fmt.Printf("  %-30s %9s %9s  %s\n", strings.Repeat("-", 30),
		strings.Repeat("-", 9), strings.Repeat("-", 9), strings.Repeat("-", 30))
	fmt.Printf("  %-30s %9s %8.2f%%  %s\n", "none (overflowing)",
		comma(int64(viewLimited.ActiveSeries())),
		100*float64(metricErrors)/float64(total), "no, and it does not say so")
	fmt.Printf("  %-30s %9s %8.2f%%  %s\n", "delete customer_id",
		comma(int64(dropped.ActiveSeries())), errRatio(dropped),
		"no, and you chose that")
	fmt.Printf("  %-30s %9s %8.2f%%  %s\n", "hash -> 16, limit 2000",
		comma(int64(hashed.ActiveSeries())), errRatio(hashed),
		"no -- still overflowing")
	fmt.Printf("  %-30s %9s %8.2f%%  %s\n", "hash -> 16, limit 50000",
		comma(int64(hashedRaised.ActiveSeries())), errRatio(hashedRaised),
		"by bucket, not by name")
	fmt.Println()
	fmt.Printf("  ground truth error ratio: %.2f%%\n", trueRatio)
	fmt.Println()
	fmt.Println("  Read the error-% column against ground truth, and read row 3")
	fmt.Println("  twice. Hashing customer_id into 16 buckets sounds like a fix and")
	fmt.Printf("  is not, because the base label set is already %s series: %s x 16\n",
		comma(int64(bounded)), comma(int64(bounded)))
	fmt.Printf("  = %s, which is %.0fx over the default limit, so it still\n",
		comma(int64(bounded)*16), float64(bounded*16)/float64(defaultCardinalityLimit))
	fmt.Println("  overflows and still lies. A BOUNDED label is not automatically a")
	fmt.Println("  SMALL one, and the default limit of 2000 is smaller than one")
	fmt.Println("  ordinary HTTP label set with anything at all added to it.")
	fmt.Println()
	fmt.Println("  Rows 2 and 4 are the two honest fixes: delete the dimension, or")
	fmt.Println("  keep a bounded version of it AND raise the limit to a number you")
	fmt.Println("  computed first. Both give the correct error ratio. Neither lets")
	fmt.Println("  you name one customer from a metric -- that capability lives in a")
	fmt.Println("  span attribute, with an exemplar as the link back. The Python")
	fmt.Println("  program in this topic covers that half.")

	fmt.Println()
	fmt.Println(strings.Repeat("=", 72))
	fmt.Println("The review question, askable in ten seconds:")
	fmt.Println("  'How many distinct values can this label take, and who decides")
	fmt.Println("   that number -- us, or a user?'")
	fmt.Println("If a user decides it, it is a span attribute, not a label.")
}

// measureBytesPerSeries builds n distinct series and measures the heap delta
// on this machine. It is a real measurement of a Go map of counters -- not of
// Prometheus, and the output says so.
func measureBytesPerSeries(n int) float64 {
	// Both readings are taken immediately after a GC, so what is compared is
	// LIVE heap and not allocation churn. Without the second GC this number
	// swings by 2x between runs, which would make it a number you should not
	// print.
	runtime.GC()
	var before, after runtime.MemStats
	runtime.ReadMemStats(&before)

	c := NewCounter(Stream{Name: "probe", AggregationCardinalityLimit: 0})
	for i := 0; i < n; i++ {
		c.Add(1, attrSet{route: "/orders", method: "GET", status: 200,
			customerID: fmt.Sprintf("cust-%08d", i)})
	}

	runtime.GC() // c is still reachable, so its map stays on the live heap
	runtime.ReadMemStats(&after)
	bytes := float64(after.HeapAlloc - before.HeapAlloc)
	runtime.KeepAlive(c)
	return bytes / float64(n)
}

func comma(n int64) string {
	s := fmt.Sprintf("%d", n)
	if n < 0 {
		return "-" + comma(-n)
	}
	var parts []string
	for len(s) > 3 {
		parts = append([]string{s[len(s)-3:]}, parts...)
		s = s[:len(s)-3]
	}
	parts = append([]string{s}, parts...)
	return strings.Join(parts, ",")
}

File: 06-observability/04-cardinality/python/cardinality_overflow.py
"""
Layer 6 Topic 4 - One label, two failures: the silent one and the loud one.

Why Python: this is the lab's own service language, and the OTel Python SDK is
where the cardinality limit is configured per-view on the MeterProvider (or
globally with OTEL_METRIC_CARDINALITY_LIMIT). The default is the spec's 2000
and there is no warning log, so you discover overflow by QUERYING for the
attribute, never by reading stderr. This program builds that behaviour out of a
dict so the mechanism is visible, then makes you look at the arithmetic that is
the only symptom.

No OpenTelemetry SDK is installed on this machine, so the counter, the
per-stream cardinality limit and the overflow attribute are ~60 lines of
standard library here. They implement the spec's rule exactly:

    "When the limit is reached, the SDK MUST record the measurement with a
     single attribute set {otel.metric.overflow: true}."

Totals stay correct. Every breakdown by the offending label starts
undercounting. Nothing errors.

What this demonstrates
----------------------
1. Series count is a PRODUCT. The napkin arithmetic, computed from the label
   set rather than asserted.
2. The silent failure: with the SDK limit in place, `sum(total)` and
   `sum by (customer_id)` disagree, and the gap is the answer to "how much of
   my dashboard is missing".
3. WHICH label values survive is first-seen-wins, so the customer who signed up
   after warm-up is invisible for the lifetime of the process.
4. Which alerts survive overflow and which quietly stop firing. This is the
   half people never check, and it is the half that pages you.
5. The loud failure: bytes per series, MEASURED on this machine with
   tracemalloc, then multiplied out to the series count the label would
   actually produce. The measured number is small; the product is not.
6. The fix, proved rather than asserted: bound the label, move the unbounded
   dimension to a span attribute, and keep the jump with an exemplar. The
   exemplar has to still get you to the trace, or the fix is not a fix.

What to look for in the output
------------------------------
Section 2's "gap" line and section 4's alert table. Everything else is
supporting arithmetic. If you read only two numbers, read the percentage of
requests that ended up in the overflow bucket and the number of alert rules
that stopped firing without changing.
"""
import random
import sys
import tracemalloc

# Spec default, per the OTel metrics SDK. Configurable per view, or globally
# with OTEL_METRIC_CARDINALITY_LIMIT.
DEFAULT_CARDINALITY_LIMIT = 2000

# 40 templated routes -- the count a real service of this size has, and the
# reason route templating is move number one in the fix list. The raw-path
# version of this list is unbounded, because `?utm_source=` is user input.
RESOURCES = ["orders", "customers", "items", "invoices", "shipments",
             "returns", "quotes", "carts"]
ROUTES = ["/health", "/ready", "/metrics", "/checkout", "/pricing/quote",
          "/search", "/login", "/logout"]
for _resource in RESOURCES:
    ROUTES += [
        "/%s" % _resource,
        "/%s/{id}" % _resource,
        "/%s/{id}/events" % _resource,
        "/%s/{id}/audit" % _resource,
    ]
METHODS = ["GET", "POST", "PUT", "DELETE", "PATCH"]
STATUSES = [200, 201, 204, 400, 401, 404, 429, 500]

CUSTOMERS = 10_000
REQUESTS = 200_000


class Counter:
    """A counter with a per-stream cardinality limit, per the OTel spec.

    Measurements past the limit are still counted -- their attributes are
    replaced by {otel.metric.overflow: True}. That is the entire behaviour, and
    it is why the totals on your dashboard stay right while the breakdowns
    stop being.
    """

    OVERFLOW = (("otel.metric.overflow", True),)

    def __init__(self, name, limit=DEFAULT_CARDINALITY_LIMIT):
        self.name = name
        self.limit = limit  # 0 or None means unlimited
        self.series = {}
        self.overflowed_measurements = 0
        self.rejected_attribute_sets = set()

    def add(self, value, **attributes):
        key = tuple(sorted(attributes.items()))
        if key in self.series:
            self.series[key] += value
            return
        unlimited = not self.limit
        # The spec reserves one slot for the overflow series itself.
        room = unlimited or len(self.series) < self.limit - 1
        if room:
            self.series[key] = value
            return
        self.series[self.OVERFLOW] = self.series.get(self.OVERFLOW, 0) + value
        self.overflowed_measurements += value
        self.rejected_attribute_sets.add(key)

    # --- the query surface, which is where the failure becomes visible ------

    def total(self):
        """sum(rate(http_server_requests_total[5m])) -- always correct."""
        return sum(self.series.values())

    def sum_by(self, label):
        """sum by (<label>) (...) -- silently loses the overflow point."""
        out = {}
        for key, value in self.series.items():
            attrs = dict(key)
            if label not in attrs:
                continue  # the overflow series has no customer_id at all
            out[attrs[label]] = out.get(attrs[label], 0) + value
        return out

    def active_series(self):
        return len(self.series)


def generate_traffic(rng, requests, customers):
    """A stream of requests. Routes are templated (bounded), customers are not."""
    for _ in range(requests):
        yield {
            "route": rng.choice(ROUTES),
            "method": rng.choices(METHODS, weights=[70, 20, 5, 3, 2])[0],
            "status": rng.choices(STATUSES, weights=[70, 8, 5, 5, 3, 4, 3, 2])[0],
            "customer_id": "cust-%05d" % rng.randrange(customers),
        }


def rule(label, width=46):
    print()
    print(label)
    print("-" * width)


def main():
    rng = random.Random(20260818)

    print("Layer 6 Topic 4 - cardinality: the silent failure, then the loud one")
    print("python %s   SDK cardinality limit default = %d attribute sets/stream"
          % (sys.version.split()[0], DEFAULT_CARDINALITY_LIMIT))
    print("=" * 72)

    # -----------------------------------------------------------------------
    rule("1. Series count is a product, and you can do it on a napkin")
    bounded = len(ROUTES) * len(METHODS) * len(STATUSES)
    print("  routes x methods x statuses      %d x %d x %d = %s series"
          % (len(ROUTES), len(METHODS), len(STATUSES), f"{bounded:,}"))
    print("  ... x customer_id                %s x %s = %s series"
          % (f"{bounded:,}", f"{CUSTOMERS:,}", f"{bounded * CUSTOMERS:,}"))
    print()
    print("  Nothing about the recording changed. Same counter, same call site,")
    print("  same request rate. One label multiplied. The code change is one word.")

    # -----------------------------------------------------------------------
    rule("2. The silent failure: SDK limit in place, dashboards undercount")

    baseline = Counter("http.server.requests")   # bounded labels only
    demo = Counter("http.server.requests")       # + customer_id, default limit

    traffic = list(generate_traffic(rng, REQUESTS, CUSTOMERS))
    for req in traffic:
        baseline.add(1, route=req["route"], method=req["method"], status=req["status"])
        demo.add(1, route=req["route"], method=req["method"], status=req["status"],
                 customer_id=req["customer_id"])

    total = demo.total()
    by_customer = demo.sum_by("customer_id")
    visible = sum(by_customer.values())
    gap = total - visible

    print("  requests recorded                %s" % f"{REQUESTS:,}")
    print("  distinct customers in traffic    %s" % f"{len({r['customer_id'] for r in traffic}):,}")
    print()
    print("  baseline counter (bounded labels)")
    print("    active series                  %s" % f"{baseline.active_series():,}")
    print("    sum(rate(...))                 %s" % f"{baseline.total():,}")
    print()
    print("  demo counter (+ customer_id, limit %d)" % demo.limit)
    print("    active series                  %s  <- capped, not grown"
          % f"{demo.active_series():,}")
    print("    sum(rate(...))                 %s  <- still correct" % f"{total:,}")
    print("    sum by (customer_id) (rate)    %s  <- what your dashboard shows"
          % f"{visible:,}")
    print("    gap                            %s  (%.1f%% of all traffic)"
          % (f"{gap:,}", 100.0 * gap / total))
    print("    customer_id values surviving   %s of %s (%.1f%%)"
          % (f"{len(by_customer):,}", f"{CUSTOMERS:,}", 100.0 * len(by_customer) / CUSTOMERS))
    print("    attribute sets rejected        %s"
          % f"{len(demo.rejected_attribute_sets):,}")
    print()
    print("  Nothing logged. Nothing errored. The only symptom is that those two")
    print("  sums disagree -- and you only ever see it if you compute both.")

    # -----------------------------------------------------------------------
    rule("3. Which values survive is first-seen-wins")
    late = Counter("http.server.requests")
    for req in traffic[:REQUESTS // 2]:
        late.add(1, route=req["route"], method=req["method"], status=req["status"],
                 customer_id=req["customer_id"])
    before = set(late.sum_by("customer_id"))

    # A new enterprise customer signs up after the process warmed up, and sends
    # a lot of traffic. Watch it not appear.
    for _ in range(5_000):
        late.add(1, route="/orders", method="GET", status=200,
                 customer_id="cust-NEW-ENTERPRISE")
    after = late.sum_by("customer_id")

    print("  customer_id values held before the new signup   %s" % f"{len(before):,}")
    print("  requests sent by cust-NEW-ENTERPRISE            5,000")
    print("  its rows on a per-customer dashboard            %s"
          % ("present" if "cust-NEW-ENTERPRISE" in after else "ABSENT"))
    print("  where those 5,000 requests went                 %s"
          % ("otel.metric.overflow=true"
             if "cust-NEW-ENTERPRISE" not in after else "its own series"))
    print()
    print("  The surviving set is whichever 1,999 attribute sets arrived first.")
    print("  That is not a sample of your customers, it is a sample of your")
    print("  start-up order, and it never refreshes while the process lives.")

    # -----------------------------------------------------------------------
    rule("4. Which alerts survive overflow, and which stop firing")

    # Ground truth, straight from the traffic we generated.
    true_5xx = sum(1 for r in traffic if r["status"] >= 500)
    true_ratio = true_5xx / REQUESTS

    # What the metric can answer. Note the trap: the overflow series carries
    # ONLY {otel.metric.overflow: true}. It has no `status` label either, so
    # any selector with a label matcher loses it too.
    metric_5xx = sum(v for k, v in demo.series.items() if dict(k).get("status", 0) >= 500)
    metric_all = demo.total()
    metric_ratio = metric_5xx / metric_all

    per_customer_visible = len(by_customer)

    print("  ground truth (from the generated traffic)")
    print("    5xx responses                  %s of %s  = %.2f%%"
          % (f"{true_5xx:,}", f"{REQUESTS:,}", 100 * true_ratio))
    print()
    print("  what each rule can still see")
    print("    %-44s %s" % ("rule", "verdict"))
    print("    %-44s %s" % ("-" * 44, "-" * 26))
    print("    %-44s intact   (%s)"
          % ("TrafficDropped: sum(rate(all)) < x", f"{metric_all:,}"))
    print("    %-44s BROKEN   (reports %.2f%%, true %.2f%%)"
          % ("ErrorRatioHigh: sum(5xx) / sum(all) > 1%", 100 * metric_ratio,
             100 * true_ratio))
    print("    %-44s BROKEN   (%s of %s customers)"
          % ("PerCustomerErrors: by (customer_id) > 5%",
             f"{per_customer_visible:,}", f"{CUSTOMERS:,}"))
    print()
    print("  The second row is the one that surprises people, and it is worth")
    print("  slowing down on. \"Totals stay correct\" means the GRAND total: the")
    print("  overflow point carries the count. It carries no other attribute, so")
    print("  the moment your query filters on a label -- status, route, method,")
    print("  anything -- the overflow point drops out of BOTH sides of your")
    print("  ratio, and the numerator loses proportionally more of it than the")
    print("  denominator does. Your error-rate alert now watches whichever 1,999")
    print("  attribute sets started first, and it will not tell you that.")
    print()
    print("  Overflow does not degrade your monitoring evenly. It degrades")
    print("  exactly the queries that name a label -- which is all the useful ones.")

    # -----------------------------------------------------------------------
    rule("5. The loud failure: what a series costs, measured then multiplied")

    def measure_series_bytes(n):
        """Build n distinct series in a fresh dict and measure the delta."""
        tracemalloc.start()
        base = tracemalloc.get_traced_memory()[0]
        counter = Counter("probe", limit=0)  # unlimited
        for i in range(n):
            counter.add(1, route="/orders", method="GET", status=200,
                        customer_id="cust-%08d" % i)
        peak = tracemalloc.get_traced_memory()[0]
        tracemalloc.stop()
        return (peak - base) / n, counter.active_series()

    per_series_10k, built_10k = measure_series_bytes(10_000)
    per_series_100k, built_100k = measure_series_bytes(100_000)
    per_series = per_series_100k

    print("  measured on THIS machine, with tracemalloc, in this process:")
    print("    %s series          %.0f bytes/series" % (f"{built_10k:>8,}", per_series_10k))
    print("    %s series          %.0f bytes/series" % (f"{built_100k:>8,}", per_series_100k))
    print()
    projected = bounded * CUSTOMERS
    print("  extrapolated (measured cost x series count -- NOT a measurement):")
    print("    %s series          %.1f GB in this toy store alone"
          % (f"{projected:>8,}", projected * per_series / 1e9))
    print()
    print("  That is one counter, in one process, with a dict for a storage")
    print("  engine. Prometheus pays more per series than this: the head block,")
    print("  the inverted index, the label-value dictionary and query fan-out")
    print("  all scale with the count. Take the shape from this number, not the")
    print("  magnitude -- and note that the shape is a straight line through the")
    print("  origin with your customer count on the x axis.")

    # -----------------------------------------------------------------------
    rule("6. The fix, and the proof that the capability survived it")

    fixed = Counter("http.server.requests")
    # A histogram bucket with an exemplar attached: a trace ID stored ALONGSIDE
    # the sample, not as a label. It creates no series, because it is not part
    # of the series identity.
    exemplars = {}  # bucket -> (trace_id, customer_id, duration)

    for i, req in enumerate(traffic):
        fixed.add(1, route=req["route"], method=req["method"], status=req["status"])
        duration = 0.02 if i % 100 else 2.4  # every 100th request is slow
        bucket = "le=0.1" if duration <= 0.1 else "le=+Inf"
        if bucket == "le=+Inf":
            exemplars[bucket] = ("%032x" % rng.getrandbits(128), req["customer_id"], duration)

    print("  1. bound the label      route templates, status classes, no raw IDs")
    print("     series after fix     %s (vs %s capped, vs %s uncapped)"
          % (f"{fixed.active_series():,}", f"{demo.active_series():,}",
             f"{bounded * CUSTOMERS:,}"))
    print("  2. move the dimension   customer_id -> span attribute (free there)")
    print("  3. keep the jump        exemplar on the slow bucket:")
    trace_id, cid, duration = exemplars["le=+Inf"]
    print("       bucket             le=+Inf")
    print("       exemplar trace_id  %s" % trace_id)
    print("       -> that trace      customer_id=%s  duration=%.2fs" % (cid, duration))
    print("       series added by the exemplar   0")
    print()
    print("  The fix is only real if the capability survives it. You still get")
    print("  from 'the p99 spiked' to one specific slow request for one specific")
    print("  customer -- in two clicks, and without buying a dimension.")

    print()
    print("=" * 72)
    print("The napkin check to run in code review, in ten seconds:")
    print("  'What is the maximum number of distinct values this label can take,")
    print("   and who decides it -- us, or a user?'")
    print("If the answer to the second half is 'a user', it is a span attribute.")


if __name__ == "__main__":
    main()
